File: cpp/Stack_and_Queue/0150_Evaluate_Reverse_Polish_Notation.cpp
#include <vector>
#include <stack>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  int evalRPN(vector <string> &tokens) {
    stack<int> stk;
    for (int i = 0; i < tokens.size(); ++i) {
      if (tokens[i] == "+" || tokens[i] == "-" || tokens[i] == "*" || tokens[i] == "/") {
        int num1 = stk.top();
        stk.pop();
        int num2 = stk.top();
        stk.pop();
        if (tokens[i] == "+") stk.push(num2 + num1);
        if (tokens[i] == "-") stk.push(num2 - num1);
        if (tokens[i] == "*") stk.push(num2 * num1);
        if (tokens[i] == "/") stk.push(num2 / num1);
      } else stk.push(stoi(tokens[i]));
    }
    return stk.top();
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: [str]) -> int:
        stk = []
        for i in tokens:
            if i not in ["+", "-", "*", "/"]:
                stk.append(i)
            else:
                if i == "+":
                    x, y = int(stk.pop()), int(stk.pop())
                    stk.append(y + x)
                elif i == "-":
                    x, y = int(stk.pop()), int(stk.pop())
                    stk.append(y - x)
                elif i == "*":
                    x, y = int(stk.pop()), int(stk.pop())
                    stk.append(y * x)
                elif i == "/":
                    x, y = int(stk.pop()), int(stk.pop())
                    stk.append(y / x)
        return stk.pop()


tokens = ["2", "1", "+", "3", "*"]
print(Solution().evalRPN(tokens))  # 9

tokens = ["4", "13", "5", "/", "+"]
print(Solution().evalRPN(tokens))  # 6

tokens = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
print(Solution().evalRPN(tokens))  # 22
